### PR TITLE
[Model] Add mistral function calling format to all models loaded with "mistral" format

### LINKED ISSUE #8301: [Bug]: Mistral Large Instruct 2407 tool calling leakage
### Your current environment

When using vllm 0.6.0, the mistral tool call parser does not work as expected for Mistral Large 2407 https://huggingface.co/mistralai/Mistral-Large-Instruct-2407 @K-Mistele 

### 🐛 Describe the bug

It used to work fine when using Autotokenizer to instantiate the tokenizer, but not with MistralTokenizer from mistral_commons.
Basically, when you run the model and give it tools, the model thinks it is the tool.
Aka, if I give Mistral Large 2407 a tool for looking up movie information on IMDB, the model responds to "Who are you" with "I am a movie database lookup bot" instead of "I am an AI trained by Mistral AI"

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by patrickvonplaten**
## Function calling for mistral models

FIX #8301

As shown in https://github.com/vllm-project/vllm/pull/8515/files#r1761725370 this PR enables function calling for all mistral models. The `tools` dict can conveniently be forwarded to `mistral_common` which takes care of correctly parsing the function calls. Tested for:
- https://huggingface.co/mistralai/Mistral-Nemo-Instruct-2407
- https://huggingface.co/mistralai/Mistral-7B-Instruct-v0.3
- https://huggingface.co/mistralai/Mistral-Large-Instruct-2407

**Make sure to load models with `--tokenizer_format="mistral"` to be certain that there are no tokenization errors**

See https://github.com/vllm-project/vllm/pull/8515/files#r1761725370 for examples.

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: examples/offline_chat_with_tools.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
+# ruff: noqa
```
Comment (patrickvonplaten) [inline examples/offline_chat_with_tools.py:1]: Example file of how to use Mistral models with function calling

**Comment by patrickvonplaten** [comment]
Gently pinging @mgoin for a review to fix #8301

File: vllm/transformers_utils/tokenizers/mistral.py [hunk 2]
```diff
@@ -165,18 +165,20 @@ def apply_chat_template(self,
...
         # encode-decode to get clean prompt
         return encoded.tokens
 
     def convert_tokens_to_string(self, tokens: List[str]) -> str:
         if isinstance(self.tokenizer, Tekkenizer):
-            return "".join(tokens)
+            return "".join([
+                t for t in tokens
+                if t not in self.tokenizer._all_special_tokens
+            ])
```
Comment (DarkLight1337) [inline vllm/transformers_utils/tokenizers/mistral.py:181]:
```suggestion
            return "".join(
                t for t in tokens
                if t not in self.tokenizer._all_special_tokens
            )
```

`str.join` works on iterables so we can avoid iterating through it twice.

**Review by DarkLight1337 — COMMENTED** [review]
Overall looks good, just a small nit which I can fix quickly.

**Review by DarkLight1337 — APPROVED** [review]

**Comment by patrickvonplaten** [comment]
@DarkLight1337 seems like mistral-nemo is too large for the tests:

```
[2024-09-17T08:37:49Z] FAILED models/decoder_only/language/test_mistral.py::test_models[5-64-bfloat16-mistralai/Mistral-Nemo-Instruct-2407] - torch.OutOfMemoryError: CUDA out of memory. Tried to allocate 140.00 MiB. GPU 0 has a total capacity of 21.99 GiB of which 112.75 MiB is free. Process 6300 has 21.85 GiB memory in use. Of the allocated memory 21.07 GiB is allocated by PyTorch, and 380.49 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)
--
  | [2024-09-17T08:37:49Z] FAILED models/decoder_only/language/test_mistral.py::test_mistral_format[5-64-bfloat16-mistralai/Mistral-Nemo-Instruct-2407] - torch.OutOfMemoryError: CUDA out of memory. Tried to allocate 60.00 MiB. GPU 0 has a total capacity of 21.99 GiB of which 20.75 MiB is free. Process 6300 has 21.94 GiB memory in use. Of the allocated memory 21.11 GiB is allocated by PyTorch, and 430.51 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)
  | [2024-09-17T08:37:49Z] FAILED models/decoder_only/language/test_mistral.py::test_mistral_function_calling[mistralai/Mistral-Nemo-Instruct-2407-bfloat16] - torch.OutOfMemoryError: CUDA out of memory. Tried to allocate 60.00 MiB. GPU 0 has a total capacity of 21.99 GiB of which 60.75 MiB is free. Process 6300 has 21.90 GiB memory in use. Of the allocated memory 21.11 GiB is allocated by PyTorch, and 390.51 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)
  | [2024-09-17T08:37:49Z] FAILED models/decoder_only/language/test_models.py::test_models[96-float-facebook/opt-125m] - RuntimeError: value cannot be converted to type at::BFloat16 without overflow
```

Should I wrap it into a "skip CI" statement?

File: tests/models/decoder_only/language/test_mistral.py [hunk 3]
```diff
@@ -81,3 +127,22 @@ def test_mistral_format(
...
+def test_mistral_function_calling(
+    vllm_runner,
+    model: str,
+    dtype: str,
+) -> None:
+    with vllm_runner(model,
+                     dtype=dtype,
+                     tokenizer_mode="mistral",
+                     config_format="mistral",
+                     load_format="mistral") as vllm_model:
```
Comment (DarkLight1337) [inline tests/models/decoder_only/language/test_mistral.py:143]: We can set `max_model_len` here to reduce the memory usage/

File: examples/offline_chat_with_tools.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+import json
+import random
+import string
+
+from vllm import LLM
+from vllm.sampling_params import SamplingParams
+
+# This script is an offline demo for running Pixtral.
```
Comment (mgoin) [inline examples/offline_chat_with_tools.py:9]: Note that this is not a demo for Pixtral

File: tests/models/decoder_only/language/test_mistral.py (hunk 3 above)
Comment (mgoin) [reply tests/models/decoder_only/language/test_mistral.py:143]: I believe this test runs on a single L4 24GB, so I think the model is just a bit too large

**Review by mgoin — APPROVED** [review]
LGTM just pending figuring the test out

File: examples/offline_chat_with_tools.py (hunk 4 above)
Comment (patrickvonplaten) [reply examples/offline_chat_with_tools.py:9]: Nice catch!

File: tests/models/decoder_only/language/test_mistral.py (hunk 3 above)
Comment (DarkLight1337) [reply tests/models/decoder_only/language/test_mistral.py:145]: I'm fine with this. Let's get it merged.

**Comment by K-Mistele** [comment]
Probably worth updating the docs with the following flags for recommended mistral tool calling:

`--tokenizer-mode mistral --load-format mistral --config-format mistral`

https://docs.vllm.ai/en/latest/serving/openai_compatible_server.html#id1

I think it's `docs/source/serving/openai_compatible_server.md`

**Comment by nithingovindugari** [comment]
Following tutorial for function calling gives me following error TypeError: LLM.chat() got an unexpected keyword argument 'tools'

**Comment by DarkLight1337** [comment]
> Following tutorial for function calling gives me following error TypeError: LLM.chat() got an unexpected keyword argument 'tools' and I have vllm version Version: 0.6.1.post2

These changes haven't been released yet - they're currently only available on `main` branch, so you have to install vLLM from source if you want to use them. By the way, make sure you're viewing the `stable` version of the docs if you're not using `main` branch of vLLM.

**Pull request merged**